path.go

```diff
@@ -139,3 +142,84 @@ func (hp *HopPayload) CountFrames() int {
 func (hp *HopPayload) CalculateRealm() {
 	hp.Realm[0] = (hp.Realm[0] & 0x0F) | (byte(hp.CountFrames()-1) << 4)
 }
+
+func (hp *HopPayload) Encode(w io.Writer) error {
+	// We'll need to add enough padding bytes to position the HMAC
+	// at the end of the payload
+	padding := hp.CountFrames()*65 - len(hp.Payload) - 1 - 32
+	if padding < 0 {
+		panic("Can't have negative padding")
+	}
+	hp.CalculateRealm()
+
+	if _, err := w.Write(hp.Realm[:]); err != nil {
+		return err
+	}
+
+	if _, err := w.Write(hp.Payload); err != nil {
+		return err
+	}
+
+	if padding > 0 {
+		w.Write(bytes.Repeat([]byte{0x00}, padding))
+	}
+
+	if _, err := w.Write(hp.HMAC[:]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (hp *HopPayload) Decode(r io.Reader) error {
+	if _, err := io.ReadFull(r, hp.Realm[:]); err != nil {
+		return err
+	}
+
+	numFrames := int(hp.Realm[0]>>4) + 1
+	numBytes := (numFrames * hopDataSize) - 32 - 1
+
+	hp.Payload = make([]byte, numBytes)
+
+	if _, err := io.ReadFull(r, hp.Payload[:]); err != nil {
+		return err
+	}
+
+	if _, err := io.ReadFull(r, hp.HMAC[:]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (hp *HopPayload) HopData() (*HopData, error) {
+	if hp.Realm[0] != 0x00 {
```

Comment:
Then how does a node extract the original hop payload when its provided with additional frames?

Reply:
Ended up modifying to make this check against the "true" realm.

Reply:
Good point, that check should have only checked the lower 4 bits, which according to the proposal are the realm.